### ISSUE TITLE
About science tools

### ISSUE THREAD

**Issue by HealthyPear** [cta-observatory/pyirf#21]
I wanted to have some input about whether or not - and if yes, how much - we should use any science tool feature _**inside**_ _pyirf_.

When I say "use", I actually want to emphasize the difference between _supporting_ a library and actually using it consistently to perform tasks (meaning making it a dependency).

I think we all agree that we need to support science tools.
(_gammapy_ and _ctools_ are the only players for now, am I wrong?)
This means that the DL3 output of _pyirf_ should be read by both in a format on which everyone agrees (which I guess is almost the case, since I think they already did some compatibility study between themselves).

What I would like to understand and trigger a conversation is whether we really need to call e.g _gammapy_ classes to compute (internally) and plot (internally or externally).
I think that performing these operations independently by any specific library would mean being less biased and / or dependent on upstream changes.

I this is not possible/advisable, we could also think of supporting multiple sources by calling their respective tools in a configurable way (and this would provide automatically a compatibility study from our side).

This discussion could be part of issues #10 and #17.

**Comment by MaxNoe** [comment]
Supporting the science tools is *easy*: we will have a well-defined format for the IRFS that every tool needs to implement. 

pyIRF needs to emit this format, the science tools need to be able to ingest said format.

So this is not really a question I think.


On the question on wether a science tool should be a dependency of this library I would tend to answer **no**, we are one level below in the stack.

The question would be what features that a science tool provides would be needed / helpful here and weigh the advantage of just using that instead of (re-)implementing that feature vs. the inverted dependency situation.

**Comment by MaxNoe** [comment]
So what features of gammapy do we currently use?

I think these are things like:

* spectral weighting
* statistical methods like calculating li & ma
* optimization / root finding

**Comment by bregeon** [comment]
I agree Max on this point... may be just adding the comment again that if we want to rely on some I/O and/or plotting libraries then I see no good reason no to rely on `ctapipe` if needed.

**Comment by HealthyPear** [comment]
> Supporting the science tools is _easy_: we will have a well-defined format for the IRFS that every tool needs to implement.
> 
> pyIRF needs to emit this format, the science tools need to be able to ingest said format.
> 
> So this is not really a question I think.

Exactly, I agree.

>
> On the question on wether a science tool should be a dependency of this library I would tend to answer **no**, we are one level below in the stack.
> 

Agree also on this

> The question would be what features that a science tool provides would be needed / helpful here and weigh the advantage of just using that instead of (re-)implementing that feature vs. the inverted dependency situation.

This was actually my point :)

For example the usage of `gammapy.spectrum.SensitivityEstimator` in `pyirf.perf.SensitivityMaker`

**Comment by MaxNoe** [comment]
There is another argument to make for not depending on science tools, at least for now.

One of the first targets was to reproduce the current IRFs from the same DL2 input. 
What if there is a difference in, e.g. the event weighting?

I think for anything directly related to calculating IRFs, we should provide our own implementation.

E.g. effective area, energy migration, sensitivity should all be implemted here to provide a reference implementation.

For other things, we might want to rely on stuff that is already there.

**Comment by MaxNoe** [comment]
For the moment, I suggest we wrap the science tools usage in our own functions. Like this, we know exactly where to change things when we decide we don't want to use certain functions of e.g. gammapy.

**Comment by HealthyPear** [comment]
Just a little update on this discussion, plus a slight variation:

- as proposed in #31 we can for sure delegate all the plotting to _gammapy_,
- even though some things, like cuts optimization, would be better implemented using an independent approach, other things, like e.g. sensitivity calculation, could depend on the particular implementation (in that case, I see no problem in implementing our own version and later add a configuration to use instead e.g. the correspondent _gammapy_ function in case it has been implemented in a different way and we want to compare the two

**Comment by HealthyPear** [comment]
I think the `pyirf.gammapy` module settles this issue - closing.

**Comment by maxnoe** [comment]
Not really. This is not about providing some interoperability (this is what `pyirf.gammapy` does) but about actually using gammapy to avoid code duplication (e.g. using `gammapy.irf` classes directly, relying on `gammapy.maps` for data storage / axes definitions or using `gammapy.models) instead of the classes in `pyirf.spectral`.
